js/body_mixin.ts

```diff
@@ -0,0 +1,230 @@
+      const enc = new TextEncoder();
+      return <ArrayBuffer>enc.encode(this.bodySource).buffer;
+    } else if (this.bodySource instanceof ReadableStream) {
+      return bufferFromStream((this.bodySource as ReadableStream).getReader());
+    } else if (this.bodySource instanceof DenoFormData) {
+      const enc = new TextEncoder();
+      return <ArrayBuffer>enc.encode(this.bodySource.toString()).buffer;
+    } else if (!this.bodySource) {
+      return new ArrayBuffer(0);
+    }
+    throw new Error(
+      `Body type not yet implemented: ${this.bodySource.constructor.name}`
+    );
+  }
+}
+
+/** @hidden */
+function validateBodyType(owner: any, bodySource: any) {
+  if (
+    bodySource instanceof Int8Array ||
+    bodySource instanceof Int16Array ||
+    bodySource instanceof Int32Array ||
+    bodySource instanceof Uint8Array ||
+    bodySource instanceof Uint16Array ||
+    bodySource instanceof Uint32Array ||
+    bodySource instanceof Uint8ClampedArray ||
+    bodySource instanceof Float32Array ||
+    bodySource instanceof Float64Array
+  ) {
+    return true;
+  } else if (bodySource instanceof ArrayBuffer) {
+    return true;
+  } else if (typeof bodySource === "string") {
+    return true;
+  } else if (bodySource instanceof ReadableStream) {
+    return true;
+  } else if (bodySource instanceof DenoFormData) {
+    return true;
+  } else if (!bodySource) {
+    return true; // null body is fine
+  }
+  throw new Error(
+    `Bad ${owner.constructor.name} body type: ${bodySource.constructor.name}`
+  );
+}
+
+export function bufferFromStream(
+  stream: ReadableStreamReader
+): Promise<ArrayBuffer> {
+  return new Promise((resolve, reject) => {
+    let parts: Array<Uint8Array> = [];
+    let encoder = new TextEncoder();
+    // recurse
+    (function pump() {
+      stream
+        .read()
+        .then(({ done, value }) => {
+          if (done) {
+            return resolve(concatenate(...parts));
+          }
+
+          if (typeof value === "string") {
+            parts.push(encoder.encode(value));
+          } else if (value instanceof ArrayBuffer) {
+            parts.push(new Uint8Array(value));
+          } else if (value instanceof Uint8Array) {
+            parts.push(value);
+          } else if (!value) {
+            // noop for undefined
+          } else {
+            console.log("unhandled type on stream read:", value);
+            reject("unhandled type on stream read");
+          }
+
+          return pump();
+        })
+        .catch(err => {
+          reject(err);
+        });
+    })();
+  });
+}
+
+/** @hidden */
+function concatenate(...arrays: Uint8Array[]): ArrayBuffer {
+  let totalLength = 0;
+  for (let arr of arrays) {
+    totalLength += arr.length;
+  }
+  let result = new Uint8Array(totalLength);
+  let offset = 0;
+  for (let arr of arrays) {
+    result.set(arr, offset);
+    offset += arr.length;
+  }
+  return <ArrayBuffer>result.buffer;
```

Comment:
don't coerce this way... 
```suggestion
  return result.buffer as ArrayBuffer;
```

But I am not sure why you even need the coercion here.